keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
+
+	// Specifies the action to be taken on the not finished (succeeded or failed) pods 
+	// when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicates that all pods wouldn't be terminated.
+	// - Terminate: indicates that not finished pods would be terminated.
+	WhenTargetAchieved WhenTargetAchievedSuccessPolicy
+}
+
+// WhenTargetAchievedSuccessPolicy specifies the action to be taken on the pods
+// when the job achieved the requirements.
+// +enum
+type WhenTargetAchievedSuccessPolicy string
+
+const (
+	// All pods wouldn't be terminated when the job reached successPolicy.
+	ContinueWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Continue"
+
+	// Not finished pods would be terminated when the job reached successPolicy.
+	TerminateWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Terminate"
+)
+```
+
+Moreover, we validate the following constraints for the `onSucceededPods`:
+- exactly one of the fields `indexes` and `count` is specified
+  for a requirement.
+- whether the specified indexes in the `indexes` and
+  the number of indexes in the `count` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `indexes` is under 64Ki.
+- exactly one of the enum `continue` or `terminate` is specified
+in the `whenTargetAchieved`.
+
+In addition, when the `successPolicy` isn't nil, we validate the `onSucceededPods` isn't nil.
+
+### Evaluation
+We evaluate the SuccessPolicy every in updating pod conditions.
+And then if the job meets the SuccessPolicy, we add `Completed` condition
+instead of `Failed` condition to `.status.conditions` even if some indexes fail.
+
+If a set of indexes are specified in the `indexes` field,
+we evaluate `.status.completedIndexes` to see if a set of indexes is there.
+
+If the number of indexes is specified in the `count`, 
+we evaluate `.status.succeeded` to see if the value is `count` or more.
+
+Also, the job controller do actions against running pods based on the value of `onSucceededPods.whenTargetAchieved`:
+
+- `continue`: Nothing. Running pods continues to run.
+- `terminate`: Not finished pods would be terminated by the job controller. Not finished pods means running or pending pods.
+
+#### The situations where successPolicy conflicts other terminating policies
+
+The successPolicy has potential conflicts with the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
+and backoffLimit in the following situations: 
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
+- when the job meets the successPolicy and the number of failed pods exceeds `backoffLimit`.
+
+So, the successPolicy will be evaluated after the pod failure policy and backoffLimit to avoid the above conflicts.
```

Comment:
That's not possible since pods can continue after the job is marked Success.  Once marked successful, the job should never be marked Failed.

I suggest adding validation to that effect.  Separate PR so we can discuss whether that's a breaking API change to prevent changes of terminal state.

Reply:
I meant that the `SuccessPolicy` isn't evaluated once the jobs are finalized by the pod failure policy and backoffLimit here. So, I guess that this has possibilities.

@deads2k Let me know if my understanding is not correct.

Reply:
That sounds like, "whichever criteria is met first wins", which I find acceptable, but needs to be clear in doc and validation must prevent the switch.

Reply:
> That sounds like, "whichever criteria is met first wins", which I find acceptable, but needs to be clear in doc

@deads2k Yes, that's right. Also, I will update this KEP for clarification and will add this notification to the website documentation.

> validation must prevent the switch.

@deads2k We should validate that the order of `successPolicies` isn't switched, right?
This means that the below `The second apply` must be rejected by validations, right?

The first apply:

```yaml
successPolicy:
    - target:
        succeededIndexes: "1-9"
        succeededCount: 5
    - target:
        succeededIndexes: "0"
        succeededCount: 1
```

The second apply:

```yaml
successPolicy:
    - target:
        succeededIndexes: "0"
        succeededCount: 1
    - target:
        succeededIndexes: "1-9"
        succeededCount: 5
```

Reply:
I made the order of the successPolicy immutable.

Reply:
The successPolicy is immutable, as well as podFailurePoliy.

Done.